aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,742 @@
+    """
+
+    hail_credits = [deepcopy(e) for e in entries]
+    for entry in hail_credits:
+
+        entry['topic'] = 'hail'
+        entry['id'] += '-credit'
+        entry['cost'] = -entry['cost']
+        entry['service'] = {
+            'id': 'aggregated-credit',
+            'description': 'Hail credit to correctly account for costs',
+        }
+        entry['sku']['id'] += '-credit'
+        entry['sku']['description'] += '-credit'
+        entry['project'] = HAIL_PROJECT_FIELD
+
+    return hail_credits
+
+
+def get_seqr_credits(entries) -> List[Dict[str, Any]]:
+    """
+    Get a hail credit for each entry
+    """
+
+    seqr_credits = [deepcopy(e) for e in entries]
+    for entry in seqr_credits:
+
+        entry['topic'] = 'seqr'
+        entry['id'] += '-credit'
+        entry['cost'] = -entry['cost']
+        entry['service'] = {
+            'id': 'aggregated-credit',
+            'description': 'Seqr credit to correctly account for costs',
+        }
+        entry['sku']['id'] += '-credit'
+        entry['sku']['description'] += '-credit'
+        entry['project'] = SEQR_PROJECT_FIELD
+
+    return seqr_credits
+
+
+async def get_batches(
+    token: str,
+    billing_project: Optional[str] = None,
+    last_batch_id: Optional[any] = None,
+):
+    """
+    Get list of batches for a billing project with no filtering.
+    (Optional): from last_batch_id
+    """
+
+    qparams = {}
+    if billing_project:
+        qparams['billing_project'] = billing_project
+
+    params = ['q=']
+    params.extend(f'q={k}:{v}' for k, v in qparams.items())
+    if last_batch_id:
+        params.append(f'last_batch_id={last_batch_id}')
+
+    q = '?' + '&'.join(params)
+    url = HAIL_BATCHES_API + q
+
+    logger.debug(f'Getting batches: {url}')
+
+    return await async_retry_transient_get_json_request(
+        url,
+        aiohttp.ClientError,
+        headers={'Authorization': 'Bearer ' + token},
+    )
+
+
+async def get_finished_batches_for_date(
+    start: datetime,
+    end: datetime,
+    token: str,
+    billing_project: Optional[str] = None,
+):
+    """
+    Get all the batches that started on {date} and are complete.
+    We assume that batches are ordered by start time, so we can stop
+    when we find a batch that started before the date.
+    """
+    batches = []
+    last_batch_id = None
+    n_requests = 0
+    skipped = 0
+
+    logger.info(f'Getting batches for range: [{start}, {end}]')
+
+    while True:
+
+        n_requests += 1
+        jresponse = await get_batches(
+            billing_project=billing_project, last_batch_id=last_batch_id, token=token
+        )
+
+        if 'last_batch_id' in jresponse and jresponse['last_batch_id'] == last_batch_id:
+            raise ValueError(
+                f'Something weird is happening with last_batch_job: {last_batch_id}'
+            )
+        last_batch_id = jresponse.get('last_batch_id')
+        if not jresponse.get('batches'):
+            return batches
```

Comment:
Add a logger statement here to say it's unusual to have no batches - and probably indicates an error